### ISSUE TITLE
Rewriting doesn’t respect eta

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3335]
It seems like rewriting doesn’t respect eta.
The last line of the code below should type check, but it gives an error instead.
```agda
{-# OPTIONS --rewriting #-}

open import Agda.Builtin.Equality
{-# BUILTIN REWRITE _≡_ #-}

record Box (P : Set) : Set where
  constructor box
  field
    unbox : P
open Box public

postulate
  A : Set
  a : A
  f : Box A → A
  f= : f (box a) ≡ a
{-# REWRITE f= #-}
  
[a] : Box A
unbox [a] = a

-- Works thanks to eta
test1 : [a] ≡ box a
test1 = refl

-- Should work as well
test2 : f [a] ≡ f (box a)
test2 = refl
```
Error:
```
f [a] != a of type A
when checking that the expression refl has type f [a] ≡ f (box a)
```

**Comment by favonia** [comment]
Could be related: #2979.

**Comment by jespercockx** [comment]
Hmm, this is a much more interesting bug than I expected since it isn't using `Prop` at all. I recently tried to make the rewriting more type-directed but it seems it's doing something wrong when it tries to match a defined symbol `f us` against a constructor pattern of an eta-record type. What it *should* do is eta-expand the value `f us`, but that's not happening somehow. Working on it...

**Comment by jespercockx** [comment]
It turned out it *was* eta-expanding `f us`, but it was also eta-expanding the record constructor pattern by applying projections to it, which is wrong since patterns don't compute. Taking the fields directly from the record constructor pattern fixes the problem.

**Comment by andreasabel** [comment]
`Whitespace violation detected in test/Succeed/Issue3335.agda.`

**Comment by jespercockx** [comment]
Oh no, I'm working on my home computer but it seems I didn't set up my environment fully.

**Comment by asr** [comment]
The example in the OP type-checks on 2.5.4.1.

**Comment by andreasabel** [comment]
@asr wrote:
> The example in the OP type-checks on 2.5.4.1.

Consequently, it is _not_ a regression released in 2.5.4.1.
